#!/usr/bin/env Rscript

# ========================================================================================
# --- Function of rehh that have been tweaked
# ========================================================================================

# ----------------------------------------------------------------------------------------
# --- ihsplot
# ----------------------------------------------------------------------------------------

ihsplot = function (ihsdata, plot.pval = TRUE, ylim.scan = 2, pch = 16,
    cex = 0.5, cex.lab = 1.25, main = NA, cex.main = 1.5, cex.axis = 1) {

    if (is.null(ihsdata$iHS) | is.null(ihsdata$frequency.class)) {
        stop("The ihsdata list is invalid and was not generated by the ihh2ihs() function...")
    }
    data = ihsdata$iHS
    lst_chrm <- unique(data$CHR)
    nbr_chrm <- length(lst_chrm)
    cum <- rep(0, (nbr_chrm + 1))
    if (nbr_chrm > 1) {
        for (i in 1:nbr_chrm) {
            cum[i + 1] <- tail(data$POSITION[data$CHR == lst_chrm[i]],
                n = 1)
        }
    }
    cum <- cumsum(cum)
    pos <- rep(0, length(data$CHR))
    pos_labels <- rep(0, nbr_chrm)
    for (i in 1:nbr_chrm) {
        pos[data$CHR == lst_chrm[i]] <- data$POSITION[data$CHR ==
            lst_chrm[i]] + cum[which(lst_chrm == lst_chrm[i])]
        pos_labels[i] <- (cum[i] + cum[i + 1])/2
    }
    col_chr <- data$CHR
    if (nbr_chrm > 1) {
        #dev.new()
        #plot.new()
        par(mar = c(5, 5, 4, 2) + 0.1)
        plot(pos, data$iHS, pch = pch, cex = cex, las = 1, col = col_chr,
            xaxt = "n", xlab = "Chromosome", ylab = expression(italic(iHS)),
            cex.lab = cex.lab, main = main, cex.main = cex.main,
            cex.axis = cex.axis)
        axis(1, at = pos_labels, labels = lst_chrm, las = 1,
            cex.lab = cex.lab, cex.axis = cex.axis)
        abline(h = c(-ylim.scan, ylim.scan), lty = 2)
        if (plot.pval) {
            #dev.new()
            #plot.new()
            par(mar = c(5, 5, 4, 2) + 0.1)
            plot(pos, data$"-log10(p-value)", pch = pch, cex = cex,
                las = 1, col = col_chr, xaxt = "n", xlab = "Chromosome",
                ylab = expression("-" * log[10] * "[1" ~ "-" ~
                  "2" ~ "|" ~ Phi[scriptstyle(italic(iHS))] ~
                  "-" ~ 0.5 * "|]"), cex.lab = cex.lab, main = main,
                cex.main = cex.main, cex.axis = cex.axis)
            axis(1, at = pos_labels, labels = lst_chrm, las = 1,
                cex.lab = cex.lab, cex.axis = cex.axis)
            abline(h = c(-ylim.scan, ylim.scan), lty = 2)
        }
    } else {
        if (max(pos) < 1000) {
            scale <- 1
            unit = "(bp)"
        } else if (max(pos) < 1e+06) {
            scale <- 1000
            unit = "(kb)"
        } else if (max(pos) < 1e+09) {
            scale <- 1e+06
            unit = "(Mb)"
        } else {
            scale <- 1e+09
            unit = "(Gb)"
        }
        #dev.new()
        #plot.new()
        par(mar = c(5, 5, 4, 2) + 0.1)
        plot(pos/scale, data$iHS, pch = pch, col = col_chr, xlab = paste("Position",
            unit), ylab = expression(italic(iHS)), cex.lab = cex.lab,
            main = main, cex.main = cex.main, cex.axis = cex.axis)
        abline(h = c(-ylim.scan, ylim.scan), lty = 2)
        if (plot.pval) {
            #dev.new()
            #plot.new()
            par(mar = c(5, 5, 4, 2) + 0.1)
            plot(pos/scale, data$"-log10(p-value)", pch = pch,
                col = col_chr, xlab = paste("Position", unit),
                ylab = expression("-" * log[10] * "[1" ~ "-" ~
                  "2" ~ "|" ~ Phi[scriptstyle(italic(iHS))] ~
                  "-" ~ 0.5 * "|]"), cex.lab = cex.lab, main = main,
                cex.main = cex.main, cex.axis = cex.axis)
            abline(h = c(-ylim.scan, ylim.scan), lty = 2)
        }
    }
    par(mar = c(5, 4, 4, 2) + 0.1)
}

# ----------------------------------------------------------------------------------------
# --- calc_ehh
# ----------------------------------------------------------------------------------------

calc_ehh = function (haplohh, mrk, limhaplo = 2, limehh = 0.05, maxgap = NA,
        plotehh = TRUE, lty = 1, lwd = 1.5, col = c("blue", "red"),
        xlab = "Position", ylab = expression(Extended ~ haplotype ~
            homozygosity ~ (italic(EHH))), cex.lab = 1.25, main = NA,
        cex.main = 1.5) {

    if (!(is.haplohh(haplohh))) {
        stop("The data are not formatted as a valid haplohh object... (see the data2haplohh() function)")
    }
    if (mrk < 1 | mrk > haplohh@nsnp) {
        stop(paste("The focal SNP index must lie between", 1,
            "and", haplohh@nsnp))
    }
    if (limhaplo < 2) {
        stop("limhaplo must be larger than 1")
    }
    if (limehh < 0 | limehh > 1) {
        stop("limehh must lie between 0 and 1")
    }
    if (is.na(maxgap)) {
        maxgap = (max(haplohh@position) + 1)
    }
    ehh <- nhaplo_eval <- matrix(0, nrow = haplohh@nsnp, ncol = 2)
    ihh <- rep(0, 2)
    res.ehh <- .C("CALL_EHH", Rdata = as.integer(haplohh@haplo),
        focal_SNP = as.integer(mrk), number_SNPs = as.integer(haplohh@nsnp),
        number_chromosomes = as.integer(haplohh@nhap), number_haplotypes = as.integer(nhaplo_eval),
        min_number_haplotypes = as.integer(limhaplo), min_EHH = as.double(limehh),
        max_gap = as.double(maxgap), map = as.double(haplohh@position),
        EHH = as.double(ehh), IHH = as.double(ihh))
    nhaplo_eval = matrix(res.ehh$number_haplotypes, 2, haplohh@nsnp,
        byrow = T)
    ehh = matrix(res.ehh$EHH, 2, haplohh@nsnp, byrow = T)
    rownames(ehh) = rownames(nhaplo_eval) = c("Ancestral allele",
        "Derived allele")
    colnames(ehh) = colnames(nhaplo_eval) = haplohh@snp.name
    ihh = res.ehh$IHH
    names(ihh) = c("Ancestral allele", "Derived allele")
    if (plotehh) {
        sel_reg <- (colSums(nhaplo_eval) > 0)
        if (sum(sel_reg) > 0) {
            if (max(haplohh@position[sel_reg]) < 1000) {
                scale <- 1
                unit = "(bp)"
            }
            else if (max(haplohh@position[sel_reg]) < 1e+06) {
                scale <- 1000
                unit = "(kb)"
            }
            else if (max(haplohh@position[sel_reg]) < 1e+09) {
                scale <- 1e+06
                unit = "(Mb)"
            }
            else {
                scale <- 1e+09
                unit = "(Gb)"
            }
            #dev.new()
            #plot.new()
            par(mar = c(5, 5, 4, 2) + 0.1)
            matplot(haplohh@position[sel_reg]/scale, t(ehh[,
                sel_reg]), col = col, type = "l", lty = lty,
                lwd = lwd, main = main, bty = "n", xlab = paste(xlab,
                  unit), ylab = ylab, cex.lab = cex.lab, cex.main = cex.main)
            abline(v = haplohh@position[mrk]/scale, lty = 2)
            if (haplohh@position[mrk] > sum(range(haplohh@position[sel_reg]))/2) {
                legend("topleft", c("Ancestral allele", "Derived allele"),
                  col = col, bty = "n", lty = lty, lwd = lwd)
            }
            else {
                legend("topright", c("Ancestral allele", "Derived allele"),
                  col = col, bty = "n", lty = lty, lwd = lwd)
            }
        }
    }
    return(list(ehh = ehh, nhaplo_eval = nhaplo_eval, freq_all1 = nhaplo_eval[1,
        mrk]/sum(nhaplo_eval[, mrk]), ihh = ihh))
}
